feat: multi version pg_net including 0.19.5

samrose opened the pull request:
## What kind of change does this PR introduce?

Build multiple versions of the pg_net extension to on different PostgreSQL versions.
Add test for the extensions and their upgrade on PostgreSQL 15 and 17. bumps version to 0.19.5 for testing as well

samrose committed:
feat: multi version pg_net including 0.19.5

samrose committed:
feat: bump ver and add tests

samrose commented:
testing pause/restore and other e2e scenarios now

samrose committed:
feat: if file exists, run prestart version switch scripts

samrose commented:
Just waiting on a branch of infra to test pause/restore dynamics on

samrose committed:
tests: bump version to test

samrose committed:
fix: rm unneeded check and fix err codes for file handling

samrose committed:
chore: bump version for testing

samrose committed:
chore: update versions for release

jfroche committed:
Move switch_pg_net_version script into separate package

jfroche committed:
Add pg-net nixos test in flake checks output

jfroche committed:
Fix shellcheck errors

jfroche committed:
feat(pg_net): switch version on file system using overlayfs

For pg_net background worker, we need to switch the version of the
extension on the file system (both pg_net.so and pg_net.control need to point
to the desired version).

As the extension is in the Nix store, we cannot simply symlink to the
desired version, as the Nix store is read-only. 
To work around this, we use overlayfs to create a writable layer on top
of the pg_net store path.

jfroche commented on nix/ext/pg_net.nix:
```diff
@@ -1,41 +1,202 @@
-    rev = "refs/tags/v${version}";
-    hash = "sha256-c1pxhTyrE5j6dY+M5eKAboQNofIORS+Dccz+7HKEKQI=";
+let
+  switchPgNetVersion = writeShellApplication {
```
I wonder if the script is trying to replace symlinks in the nix store ? Maybe we copy postgres lib/extension directory before ?

samrose replied:
@jfroche good call. I talked with @jfroche today and he's going to propose a better way to do this for the short term for all packages with background worker.

jfroche replied:
I have added a change for pg_net. We will be able to refactor for the next extension using a background worker.

samrose committed:
feat:
limit using verison in extension creation to admin user for all extensions
by using supautils and before-create hooks
add test for this in pg_net

samrose committed:
Revert "feat:"

This reverts commit 268cadb9eab2b858a376775a9ae086a7be99fe15.

samrose committed:
fix: fmt

samrose committed:
chore: bump version for testin

steve-chavez commented on nix/ext/pg_net.nix:
```diff
@@ -1,41 +1,224 @@
+      buildInputs = [
+        curl
+        postgresql
+      ] ++ lib.optional (version == "0.6") libuv;
```
Q: Is it necessary to support such and old version of the extension? The `libuv` version was not stable and I'm not sure if it was actually deployed on Supabase. Or if any project with such an old version survives.

samrose replied:
@steve-chavez we'll take a look again at the projects, and if we don't need this we could drop it. Thanks!

za-arthur commented on nix/ext/pg_net.nix:
```diff
@@ -1,41 +1,224 @@
+
+        # Create versioned control file with modified module path
+        sed -e "/^default_version =/d" \
+            -e "s|^module_pathname = .*|module_pathname = '\$libdir/${pname}'|" \
```
This will not set `module_pathname` since `pg_net` itself doesn't set it:
https://github.com/supabase/pg_net/blob/master/pg_net.control.in

samrose replied:
@za-arthur great catch and thank you! 

I am attempting to resolve it with https://github.com/supabase/postgres/pull/1744/commits/4996d81a8596ffbb903a8eceb2b1c8d0e153bc7c and will test and verify that that approach does everything we need. Feel free to continue to feedback on this of course.

samrose replied:
I was wrong: https://github.com/supabase/postgres/commit/4996d81a8596ffbb903a8eceb2b1c8d0e153bc7c  won't make a difference because the db will have a value of `pg_net` for `probin` however, the script that creates aliases should solve this issue for any extension with this problem, and would need that coverage even if the extensions are fixed upstream (due to values in existing databases)

samrose replied:
...reverted with https://github.com/supabase/postgres/pull/1744/commits/381035f287a85287f94c63f517be230058ef30ac

samrose committed:
fix: control files need module_pathname

samrose committed:
Revert "fix: control files need module_pathname"

This reverts commit 4996d81a8596ffbb903a8eceb2b1c8d0e153bc7c.

This change would not resolve the issue it was targeting.

samrose committed:
chore: version bump for testing

steve-chavez commented on nix/ext/pg_net.nix:
```diff
@@ -1,41 +1,224 @@
+          rm sql/pg_net--0.5.1--0.6.sql
+        '';
+
+      env.NIX_CFLAGS_COMPILE = "-Wno-error";
```
Shouldn't be needed starting from 0.19.1, check
https://github.com/NixOS/nixpkgs/pull/425390/files

samrose replied:
resolved in https://github.com/supabase/postgres/pull/1744/commits/146bd2787777e185417f488378b8c5371d22d4cd

valigula reviewed (approved):
LGTM

samrose committed:
fix: limit flag to versions less than 0.19.1

samrose committed:
chore: bump version for release
